chaintest/action_test_helpers.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package chaintest
+
+import (
+	"context"
+	"testing"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ava-labs/hypersdk/chain"
+	"github.com/ava-labs/hypersdk/codec"
+	"github.com/ava-labs/hypersdk/state"
+)
+
+var _ state.Mutable = (*InMemoryStore)(nil)
+
+// InMemoryStore is a storage that acts as a wrapper around a map and implements state.Mutable.
+type InMemoryStore struct {
+	Storage map[string][]byte
+}
+
+func NewInMemoryStore() *InMemoryStore {
+	return &InMemoryStore{
+		Storage: make(map[string][]byte),
+	}
+}
+
+func (i *InMemoryStore) GetValue(_ context.Context, key []byte) ([]byte, error) {
+	val, ok := i.Storage[string(key)]
+	if !ok {
+		return nil, database.ErrNotFound
+	}
+	return val, nil
+}
+
+func (i *InMemoryStore) Insert(_ context.Context, key []byte, value []byte) error {
+	i.Storage[string(key)] = value
+	return nil
+}
+
+func (i *InMemoryStore) Remove(_ context.Context, key []byte) error {
+	delete(i.Storage, string(key))
+	return nil
+}
+
+// ActionTest is a single parameterized test. It calls Execute on the action with the passed parameters
+// and checks that all assertions pass.
+type ActionTest struct {
+	Name string
+
+	Action chain.Action
+
+	Rules     chain.Rules
+	State     state.Mutable
+	Timestamp int64
+	Actor     codec.Address
+	ActionID  ids.ID
+
+	ExpectedOutputs [][]byte
+	ExpectedErr     error
+
+	Assertion func(state.Mutable) bool
+}
+
+// Run execute all tests from the test suite and make sure all assertions pass.
+func Run(t *testing.T, tests []ActionTest) {
+	for _, test := range tests {
+		t.Run(test.Name, func(t *testing.T) {
+			require := require.New(t)
+
+			output, err := test.Action.Execute(context.TODO(), test.Rules, test.State, test.Timestamp, test.Actor, test.ActionID)
```

Comment:
Thanks! I have replace the `context.TODO` by a `context.Background` and the test now checks that the `Err()` is not `nil` by using `require.NoError(ctx.Err())`. Is that what you intended ?

Reply:
Though if the context is passed from the `Run`, is that really okay to check for the done signal for each of the subtest ? Wouldn't it be better to send a fatal error so that it completely stops the test suite ? because the context is going to be passed to the subsequent call to `Execute` which would already be populated in the done channel ?
I'm wondering if the context shouldn't be created for each subtest again here

Reply:
> Thanks for the litterature!

No problem

> I have replaced the context.TODO by a context.Background

The important thing is that this happens as far "out" as possible, i.e. the `TestXXX()` function (as you've done). The idea being that the users of the suite can supply their own.

> and the test now checks that the Err() is not nil by using require.NoError(ctx.Err()). Is that what you intended ?

No, you can completely ignore `ctx.Err()` and probably should as it's feasible that a test may want to test for this error (i.e. an error is expected).